### ISSUE TITLE
Option to handle dependency

### ISSUE THREAD

**Issue by testn** [rhuss/docker-maven-plugin#326]
It would be nice to handle the jar dependencies apart from your own jars when building docker. The benefit of doing so is to reuse the docker caching layer as much as possible. This might however break the compatbility with assembly descriptor file. What do you think?

**Comment by jgangemi** [comment]
could you provide a little more information on your use case? i'm not sure i fully understand it.

**Comment by testn** [comment]
most of the times, the dependencies do not change. So if you provide an ability to build the jar dependency layer separating from app jar itself, it could save the time to build the docker image significantly.

**Comment by rhuss** [comment]
If I understand you right, it's about an extra image for stuff which doesn't change much, putting on top a single image layer with the compiled app jar (so that not all dependencies need to be copied over and transmitted to the Docker daemon for each run).

Sounds quite useful but is probably a bit complicated to integrate, partly because the configuration structure would need to change (i.e. a plain assembly descriptor is not good enough) and partly because one need to detect, too, when this 'background' image layer has been changed somehow (so involving some state management). 

Maybe we extend the assembly descriptor xsd with some sort of annotation attributes to achieve this.

Alternatively, when we better support plain Dockerfile syntax with expression expansion we could add there support for multiple docker files for a single build, one for holding the dependencies (which are referenced from within this Dockerfile with a special property) and one for the artifact to build, inheriting from the first. Then, one would need to detect when the first image needs to be rebuild. 

Or one could use a single Dockerfile with extra annotations (maybe via Dockerfile comments ?) marking certain parts as 'static', so they are extracted and introduced in a base image.

Still in brainstorming mode ;-) ....

**Comment by jgangemi** [comment]
why not just build a 'seed' container , controlled by a profile, that has these dependencies inside of it and have the application container inherit from it?

**Comment by chris-snider** [comment]
We do something similar using a Dockerfile and Maven executions.

Setup the pom to copy the target dependencies to a special folder such as ${project.build.dir}/dep_source.  Then in the Dockerfile, closer to the top, add  ADD/COPY command to copy the source /dep_source /image/path/final/destination/.

Lastly, to force docker to invalidate the caches, add an environment variable ENV FORCE_BUILD=1 at a high level in the Dockerfile allows for invalidating the cache and every layer below it.  If Docker doesn’t catch the difference when the dependencies *do* change, changing the environment variable invalidation for that layer and all subsequent layers.


Chris Snider
Senior Software Engineer
Intelligent Software Solutions, Inc.
[Description: Description: Description: cid:image001.png@01CA1F1F.CBC93990]

**Comment by rhuss** [comment]
@jgangemi sounds a good solution which will work with the current setup. Maybe we should start a collection of sample project which highlights this sort of build patterns ? 

@chris-snider good idea for taking benefit of caching layers. However, for the build the dependencies' jars still has to be send to the Docker daemon in a `build.tar` a long with the Dockerfile (even when then cached by the Docker daemon), so its not the optimal solution in this case when the dependency set is large.

**Comment by jgangemi** [comment]
@rhuss yeah - a cookbook full of pom configs or links that point to various gists.
